test/e2e/daemon_set.go

```diff
@@ -357,40 +356,54 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
 		Expect(err).NotTo(HaveOccurred())
 		Expect(newDS.Spec.TemplateGeneration).To(Equal(templateGeneration))
 
+		By(fmt.Sprintf("Wait for all pods to be adopted by DaemonSet %s", newDSName))
+		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkDaemonSetPodsAdopted(c, ns, newDS.UID, label))
+		Expect(err).NotTo(HaveOccurred(), "error waiting for DaemonSet pods to be orphaned")
+
 		By(fmt.Sprintf("Make sure no daemon pod updated its template generation %d", templateGeneration))
 		err = checkDaemonPodsTemplateGeneration(c, ns, label, fmt.Sprint(templateGeneration))
 		Expect(err).NotTo(HaveOccurred())
 
-		By("Make sure no pods are recreated")
-		newDSPodsFirstCreationTime := getDaemonPodsFirstCreationTime(c, ns, label)
-		Expect(newDSPodsFirstCreationTime.Before(dsPodsLastCreationTime) ||
-			newDSPodsFirstCreationTime.Equal(dsPodsLastCreationTime)).To(BeTrue())
+		By("Make sure no pods are recreated by looking at their names")
+		err = checkDaemonSetPodsName(c, ns, dsName, label)
+		Expect(err).NotTo(HaveOccurred())
 
+		// 4. Orphan DaemonSet pods again
 		By(fmt.Sprintf("Deleting DaemonSet %s and orphaning its pods", newDSName))
-		orphanDependents := true
-		err = c.Extensions().DaemonSets(ns).Delete(newDSName, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
+		err = orphanDaemonSetPods(c, newDS)
 		Expect(err).NotTo(HaveOccurred())
+		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkDaemonSetPodsOrphaned(c, ns, label))
+		Expect(err).NotTo(HaveOccurred(), "error waiting for DaemonSet pods to be orphaned")
 		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkDaemonSetDeleted(f, ns, newDSName))
 		Expect(err).NotTo(HaveOccurred(), "error waiting for DaemonSet to be deleted")
 
-		newRestartDSName := dsName + "-new-restart"
+		// 4. Adopt DaemonSet pods (should kill and restart those pods)
+		newRestartDSName := "restart"
 		By(fmt.Sprintf("Creating a new RollingUpdate DaemonSet %s to restart adopted pods", newRestartDSName))
 		newRestartDS := newDaemonSet(newRestartDSName, image, label)
 		newRestartDS.Spec.UpdateStrategy = extensions.DaemonSetUpdateStrategy{Type: extensions.RollingUpdateDaemonSetStrategyType}
 		newRestartDS, err = c.Extensions().DaemonSets(ns).Create(newRestartDS)
 		Expect(err).NotTo(HaveOccurred())
 		Expect(newRestartDS.Spec.TemplateGeneration).To(Equal(int64(1)))
 
-		By("Wait for all DaemonSet pods template Generation to be updated to 1")
-		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, templateGenerationMatch(c, ns, label, "1"))
-		Expect(err).NotTo(HaveOccurred(), "error waiting for daemon pod template generation to be 1")
+		By("Wait for restarted DaemonSet pods launch on every node of the cluster.")
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
I see that you meant this `Expect(err).NotTo(HaveOccurred())` error check is redundant. Removed. I also removed other redundant error checks and add a missing error check.